### PR TITLE
build script improvements

### PR DESCRIPTION
**Pull request by edgomez**
I think I addressed quite a lot of shortcomings in the build script. It may be worth merging this PR even if this touches a script that should not receive much attention after all. Dt is all about the pics, not its build system.

quantity of changes might be frightening, but overall result is a much cleaner, saner, and safer build script.

Maintenability work:
- features are declared at the beginning of the script, this should help avoid devs being to add support for their cool new cmake USE_{FEATURE} option at 3 different places in the script
- move things in functions to ease script reading

New feats:
- added option to enable address sanitizer builds easily (for devs)
- added support --build-generator, now Ninja can be used easily, we can check that our CMake rules do not expect cmake+make specific behavior
- added support for --install, that installs after the build
- added support for --sudo, that prefixes install commands with sudo to cover the advertised default install path /opt (personnaly i use ~/opt to avoid installing as root)
- added --skip-config, that just prints the cmake command line that would have been executed, should help people that would like to fine tune something that is not covered by the script options  maybe

Fixes and cleanup:
- use cmake --build <dir> --target <whatever> -- -j<parallel> to actually build the binaries, this avoids having to deal explicitly with the chosen build program (make, gmake, ninja, Eclipse CDT ...)
- set -e earlier to avoid any possible error
- support build paths with missing intermediary dirs
- prefer $() instead of ``, it is usually more reliable
- updated help message that was missing libsecret for example

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -1,191 +1,215 @@
...
-   --enable-kwallet
-   --enable-openmp
-   --enable-opencl
-   --enable-unity
-   --enable-tethering
-   --enable-geo
-   --enable-lua
-   --enable-openexr
-   --enable-webp
+By default cmake will enabel the features it autodetects on the build machine.
```
Comment (boucman) [inline build.sh]: enable

File: build.sh [hunk 2]
```diff
@@ -241,19 +250,31 @@ if [ ! -d "$BUILD_DIR" ]; then
...
-&& $MAKE -j $MAKE_TASKS
+	\"$DT_SRC_DIR\" 
+cd "$OLDPWD"
+cmake --build "$BUILD_DIR" -- -j$MAKE_TASKS
 
 if [ $? = 0 ]; then
 	cat <<EOF
 Darktable finished building, to actually install darktable you need to type:
-# cd "$BUILD_DIR"; sudo make install
+\$ cmake --build "$BUILD_DIR" --target install # optionnaly prefixed by sudo
```
Comment (christte) [inline build.sh]: optionally

File: build.sh [hunk 3]
```diff
@@ -126,9 +135,11 @@ Installation:
...
 
 Build:
-   --builddir <string>   Building directory (default: $DT_SRC_DIR/build)
-   --buildtype <string>  Build type (Release, Debug, default: Release)
--j --jobs <integer>      Number of tasks (default: number of CPUs)
+   --builddir <string>       Building directory (default: $DT_SRC_DIR/build)
+   --buildtype <string>      Build type (Release, Debug, RelWithDebInfo)
+                                        (default:RelWithDebInfo)
+   --buildgenerator <string> Build tool (default: Unix Makefiles)
+-j --jobs <integer>          Number of tasks (default: number of CPUs)
```
Comment (christte) [inline build.sh]: Should --asan be listed here?

File: build.sh [hunk 4]
```diff
@@ -39,38 +40,8 @@ parse_feature()
...
-	geo)
-		OPT_GEO=$value
-		;;
-	openexr)
-		OPT_OPENEXR=$value
-		;;
-	webp)
-		OPT_WEBP=$value
+	flickr|libsecret|kwallet|openmp|opencl|unity|tethering|lua|geo|openexr|webp)
+		eval "OPT_$(printf $feature|tr a-z A-Z)"=$value
```
Comment (christte) [inline build.sh]: neat!

File: build.sh [hunk 5]
```diff
@@ -1,191 +1,215 @@
...
-			if [ "$?" = "0" ]; then
-				MAKE_TASKS=$TMP_CORES
+			ncpu=$(/sbin/sysctl -n hw.ncpu 2>/dev/null)
+			if [ $? -neq 0 ]; then
+				ncpu=-1
 			fi
 		fi
+		;;
+	*)
+		printf "warning: unable to determine number of CPU on $platform\n"
```
Comment (christte) [inline build.sh]: "number of CPUs" ?

File: build.sh [hunk 6]
```diff
@@ -225,36 +256,71 @@ CMAKE_MORE_OPTIONS="${CMAKE_MORE_OPTIONS} ${CMAKE_PREFIX_PATH}"
...
-	-DCMAKE_INSTALL_PREFIX=${INSTALL_PREFIX} \
-	-DCMAKE_BUILD_TYPE=${BUILD_TYPE} \
-	${CMAKE_MORE_OPTIONS} \
-	"$DT_SRC_DIR" \
-&& $MAKE -j $MAKE_TASKS
 
-if [ $? = 0 ]; then
+if [ $DO_BUILD -eq 0 ] ; then
+	cat <<EOF
+Darktable configuration is finished, to actually build and install darktable
```
Comment (christte) [inline build.sh]: Would this be more readable presented on two lines? Also, darktable is always written with lowecase D.

> The darktable configuration is finished.
> To build and install darktable you need to type:

File: build.sh [hunk 7]
```diff
@@ -225,36 +256,71 @@ CMAKE_MORE_OPTIONS="${CMAKE_MORE_OPTIONS} ${CMAKE_PREFIX_PATH}"
...
+EOF
+	exit 0
+fi
+
+# build the binaries
+eval "$cmd_build"
+
+if [ $DO_INSTALL -eq 0 ] ; then
 	cat <<EOF
 Darktable finished building, to actually install darktable you need to type:
```
Comment (christte) [inline build.sh]: Suggestion: Separate sentences, lowercase D.

darktable finished building. To actually install darktable you need to type:

Comment (edgomez) [inline build.sh]: addressed in 331aa0e

File: build.sh (hunk 6 above)
Comment (edgomez) [inline build.sh]: addressed in 331aa0e

File: build.sh (hunk 5 above)
Comment (edgomez) [inline build.sh]: addressed in 331aa0e

File: build.sh (hunk 1 above)
Comment (edgomez) [inline build.sh]: addressed in 331aa0e

**Comment by edgomez** [comment]
I rebased on top of current master.

File: build.sh [hunk 8]
```diff
@@ -204,18 +228,25 @@ cmake_boolean_option()
...
+
+features_set_to_autodetect
+parse_args "$@"
+
+if [ $PRINT_HELP -ne 0 ] ; then
+	print_help
+	exit 1
+fi
+
+MAKE_TASKS=$(num_cpu)
```
Comment (christte) [inline build.sh]: This assignment overrides the value set by parse_args, causing -j to be ignored.

**Comment by christte** [comment]
Thanks for the contribution!

File: build.sh (hunk 8 above)
Comment (edgomez) [inline build.sh]: Oops i didn't see that comment in my email.

It was easier to just move the MAKE_vars lines above features_set_to_autodetect call.

Too bad you merged fast. I'll just modify directly the script as it's a very small change, it makes the code more straightforward

Comment (christte) [inline build.sh]: Not sure if I get what you mean in your last comment :-) I've fixed it in master the way I sughested above, but feel free to share additional fixes or suggestions.

**Pull request merged**